python/pyspark/sql/datasource.py

```diff
@@ -469,6 +501,188 @@ def stop(self) -> None:
         ...
 
 
+class SimpleInputPartition(InputPartition):
+    def __init__(self, start: dict, end: dict):
+        self.start = start
+        self.end = end
+
+
+class SimpleDataSourceStreamReader(ABC):
+    """
+    A base class for simplified streaming data source readers. Compared to DataSourceStreamReader,
+    SimpleDataSourceStreamReader doesn't require planning data partitioning. Also, the read api of
+    SimpleDataSourceStreamReader allows reading data and planning the latest offset at the same time.
+
+    .. versionadded: 4.0.0
+    """
+
+    def initialOffset(self) -> dict:
+        """
+        Return the initial offset of the streaming data source.
+        A new streaming query starts reading data from the initial offset.
+        If Spark is restarting an existing query, it will restart from the check-pointed offset
+        rather than the initial one.
+
+        Returns
+        -------
+        dict
+            A dict or recursive dict whose key and value are primitive types, which includes
+            Integer, String and Boolean.
+
+        Examples
+        --------
+        >>> def initialOffset(self):
+        ...     return {"parititon-1": {"index": 3, "closed": True}, "partition-2": {"index": 5}}
+        """
+        raise PySparkNotImplementedError(
+            error_class="NOT_IMPLEMENTED",
+            message_parameters={"feature": "initialOffset"},
+        )
+
+    def read(self, start: dict) -> (Iterator[Tuple], dict):
+        """
+        Read all available data from specified start offset and return the offset that next read attempt
+        starts from.
+
+        Parameters
+        ----------
+        start : dict
+            The start offset to start reading from.
+
+        Returns
+        -------
+        A tuple of an iterator of :class:`Tuple` and a dict\\s
+            The iterator contains all the available records after start offset.
+            The dict is the end of this read attempt and the start of next read attempt.
+        """
+        raise PySparkNotImplementedError(
+            error_class="NOT_IMPLEMENTED",
+            message_parameters={"feature": "read"},
+        )
+
+    def read2(self, start: dict, end: dict) -> Iterator[Tuple]:
```

Comment:
can we have one method? you can make `end` argument optional

Reply:
They are fundamentally different, the former read() is to read data and plan end offset, the latter is to read data between already planned start and end offset.

Reply:
can we have a different name in that case instead of `read2`

Reply:
Changed to readBetweenOffsets()

Reply:
cc @HeartSaVioR

Reply:
I have less context so let's make sure this is signed off by @HeartSaVioR

Reply:
The name itself might be OK. Maybe we have an option to make the both of method names be self-descriptive (not just read), but if we prefer shorter name, maybe OK to have either to be "read".

I see a bigger issue on implementation. Let's address that first.

Reply:
There can't be 2 methods named read() for the same class, python doesn't have method overloading IIRC.

Reply:
It can't have overloaded ones but it can dispatch by embedding if-else, and leveraging optional argument. e.g.,

```python
def read(self, start: dict, end: dict = None) -> Union[Tuple[Iterator[Tuple], dict], Iterator[Tuple]:

    if end is None:
        return # logic for read(start)
    else:
        return # logic for read(start, end)
```

Reply:
Then we won't be able to enforce that read with end offset is implemented.

Reply:
Yeah that would have to be documented. BTW, in Python you can't enforce anything in any event.